ruby/ql/docs/flow_summaries.md

```diff
@@ -0,0 +1,205 @@
+```ql
+private class ChompSummary extends SimpleSummarizedCallable {
+  ChompSummary() { this = "chomp" }
+
+  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+    input = "Argument[self]" and
+    output = "ReturnValue" and
+    preservesValue = false
+  }
+}
+```
+
+The shared dataflow library will use this summary to construct a fake definition
+for `chomp`. The behaviour of this definition depends on the body of
+`propagatesFlowExt`. In this case, the method will propagate taint flow from the
+`self` argument (i.e. the receiver) to the return value.
+
+If `preservesValue = true` then value flow is propagated. If it is `false` then
+only taint flow is propagated.
+
+Any call to `chomp` in the database will be translated, in the dataflow graph,
+to a call to this fake definition. 
+
+`input` and `output` define the "from" and "to" locations in the flow summary.
+They use a custom string-based syntax which is similar but not identical to
+Models as Data. These strings are often referred to as access paths.
+
+# Syntax
+
+Access paths consist of zero or more components separated by dots (`.`). The
+permitted components differ for input and output paths. The meaning of each
+component is defined relative to the implicit context of the component as
+defined by the preceding access path. For example,
+
+```
+Argument[0].Element[1].ReturnValue
+```
+
+refers to the return value of the element at index 1 in the array at argument 0
+of the method call.
+
+## `Argument` and `Parameter`
+
+The `Argument` and `Parameter` components refer respectively to an argument to a
+call or a parameter of a callable. They contain one or more _specifiers_[^1] which
+constrain the range of arguments/parameters that the component refers to. For
+example, `Argument[0]` refers to the first argument.
+
+If multiple specifiers are given then the result is a disjunction, meaning that
+the component refers to any argument/parameter that satisfies at least one of
+the specifiers. For example, `Argument[0, 1]` refers to the first and second
+arguments.
+
+### Specifiers
+
+#### `self`
+The receiver of the call.
+
+#### `<integer>`
+The argument to the method call at the position given by the integer. For
+example, `Argument[0]` refers to the first argument to the call.
+
+#### `<integer>..`
+An argument to the call at a position greater or equal to the integer. For
+example, `Argument[1..]` refers to all arguments except the first one. This
+specifier is not available on `Parameter` components.
+
+#### `<string>:`
+A keyword argument to the call with the given name. For example,
+`Argument[foo:]` refers to the keyword argument `foo:` in the call.
+
+#### `block`
+The block argument passed to the call, if any.
+
+#### `any`
+Any argument to the call, except `self` or `block` arguments.
+
+#### `any-named`
+Any keyword argument to the call.
+
+#### `hash-splat`
+The special "hash splat" argument/parameter, which is written as `**args`.
```

Comment:
Note that this will also include all explicit keyword arguments, wrapped in an implicit hash splat argument.